File: resources/asteroids/scripts/Rock.js
import("Moving.js")

Rock = Class({
    Extends: Moving,
    
    init: function() {
    	Super();
        this.rotationSpeed = 0;
        this.hits = 0;
    },
    
    onBirth: function() {
        this.role.addTag("shootable");
        this.role.addTag("deadly");
        sceneDirectorScript.addRocks(1);
    },
    
    tick: function() {
    	Super();
    	
        this.actor.getAppearance().adjustDirection(this.rotationSpeed);
    },
    
    shot: function(bullet) {
    	// Small bullets have NO effect on strong rocks.
    	if (bullet.getCostumeProperties().strength < this.getCostumeProperties().strength) {
    		this.actor.event("ricochet");
    		return;
    	}
    	
    	this.hits += bullet.getCostumeProperties().strength;
    	// Have we hit the rock enough times?
    	if ( this.hits < this.getCostumeProperties().hitsRequired) {
    		this.actor.event("hit");
    		return;
    	}
    	
		this.actor.event("explode");
    	directorScript.addPoints(this.getCostumeProperties().points);
    	
        new itchy.extras.Explosion(this.actor)
            .spread( bullet.actor.getHeading() - 120, bullet.actor.getHeading() + 120 ).randomSpread()
            .speed(5,3,0,0).fade(3).distance(40)
            .rotate(true).pose("fragment").projectiles(8	)
            .createActor();

        var pieces = this.getCostumeProperties().pieces;
        for (var i = 0; i < pieces; i ++ ) {
            var actor = this.actor.createCompanion("fragment");
            var role = actor.getRole().roleScript;
            
            role.rotationSpeed = 3 * i - this.rotationSpeed;
            role.vx = itchy.util.Util.randomBetween( -4, 4 );
            role.vy = itchy.util.Util.randomBetween( -4, 4 );
            
            actor.setDirection( this.actor.getHeading() + i * 120 );
            actor.moveForwards( 40 );
        }
        sceneDirectorScript.addRocks(-1);
        this.actor.deathEvent("explode");
        this.role.removeTag("shootable");
    }
});
RoleScript.addProperty("Rock", "rotationSpeed", Double, "Rotation Speed (Degrees per Tick)");
RoleScript.addProperty("Rock", "vx", Double, "X Velocity");
RoleScript.addProperty("Rock", "vy", Double, "Y Velocity");

File: resources/Space-Miner/scripts/rock.py
from common import *

from moving import Moving

properties = ArrayList()
properties.add( DoubleProperty("strength") )
properties.add( DoubleProperty("rotationSpeed").hint("Degrees per Tick)") )
properties.add( DoubleProperty("vx").label("X Velocity") )
properties.add( DoubleProperty("vy").label("Y Velocity") )

game = Itchy.getGame()
director = game.getDirector()

class Rock(Moving) :

    def __init__(self) :
        Moving.__init__(self)
        self.strength = 1
        self.rotationSpeed = 0
        self.hits = 0
        
        
    def onBirth(self):
        game.getSceneDirector().addRocks(1)


    def onAttach(self):
        Moving.onAttach(self)
        self.addTag("shootable")
        self.addTag("deadly")

        
    def tick(self):
        Moving.tick(self)
        self.getActor().getAppearance().adjustDirection(self.rotationSpeed)

    
    def shot(self, bullet) :
    	# Small bullets have NO effect on strong rocks.
    	strength = bullet.getActor().getCostume().getProperties().strength
    	
    	if (strength < self.getActor().getCostume().getProperties().strength) :
    		self.getActor().event("ricochet")
    		return
    	
    	self.hits += strength
    	# Have we hit the rock enough times?
    	if self.hits < self.getActor().getCostume().getProperties().hitsRequired :
    		self.getActor().event("hit")
    		return
    	
		self.getActor().event("explode")
    	director.addPoints(self.getActor().getCostume().getProperties().points)
    	
        ExplosionBuilder(self.getActor()) \
            .spread( bullet.getActor().getHeading() - 120, bullet.getActor().getHeading() + 120 ).randomSpread() \
            .speed(5,3,0,0).fade(3).distance(40) \
            .rotate(True).pose("fragment").projectiles(8) \
            .create()

        sum_dx = 0
        sum_dy = 0
        pieces = self.getActor().getCostume().getProperties().pieces
        for i in range( 0, pieces ) :
            actor = self.getActor().createCompanion("fragment-"+`i + 1`)
            role = actor.getRole()
            
            explosiveness = bullet.getActor().getCostume().getProperties().explosiveness
            role.rotationSpeed = self.rotationSpeed + Util.randomBetween( -explosiveness, explosiveness )

            if i == pieces -1 :
                dx = -sum_dx
                dy = -sum_dy
            else :
                dx = Util.randomBetween( -explosiveness, explosiveness )
                dy = Util.randomBetween( -explosiveness, explosiveness )
                sum_dx += dx
                sum_dy += dy
            role.vx = self.vx + dx
            role.vy = self.vy + dy
            
            actor.setDirection( self.getActor().getAppearance().getDirection() )

        game.getSceneDirector().addRocks(-1)
        self.getActor().deathEvent("explode")
        self.removeTag("shootable")
    
    
    # Boiler plate code - no need to change this
    def getProperties(self):
        return properties

    # Boiler plate code - no need to change this
    def getClassName(self):
        return ClassName( Role, self.__module__ + ".py" )


